### PR TITLE
feat(listenbrainz): init

### PR DESCRIPTION
**Pull request by 00dani**
## 🎉 Theme for [ListenBrainz](https://listenbrainz.org) 🎉

ListenBrainz keeps track of music you listen to and provides you with insights into your listening habits. Your listens are associated with music information in the MusicBrainz database to provide more detailed and accurate song metadata, either automatically or through manual song mapping.

preview

## 💬 Additional Comments 💬

I've thoroughly themed the parts of ListenBrainz I use personally, making sure they look great with all four Catppuccin flavours. However I prefer to use a local music player rather than its built-in player feature, so that area got less attention - I've attempted to make this look nice as well, but I don't know how it behaves with every possible music service.

It's also possible I've entirely missed parts of how ListenBrainz works simply because I haven't been using it all that long. Updates to this style are very welcome! Asking me to look into it is also fine. 👍

## 🗒 Checklist 🗒

- [x] I have read and followed Catppuccin's [submission guidelines](https://github.com/catppuccin/userstyles/blob/main/docs/userstyle-creation.md).
- [x] I have made a new directory underneath `/styles/<name-of-website>` containing the contents of the [`/template`](https://github.com/catppuccin/userstyles/blob/main/template/) directory.
  - [x] I have ensured that the new directory is in **lower-kebab-case**.
  - [x] I have followed the template and kept the preprocessor as [LESS](https://lesscss.org/#overview).
- [x] I have made sure to update the
      [`userstyles.yml`](https://github.com/catppuccin/userstyles/blob/main/scripts/userstyles.yml)
      file with information about the new userstyle.
- [x] I have included the following files:
  - `catppuccin.user.css` - all the CSS for the userstyle, based on the
    template.
  - `preview.webp` - composite image of all four individual flavor screenshots stitched together,
    generated via [Catwalk](https://github.com/catppuccin/toolbox/tree/main/catwalk#readme).

### CODE REVIEW

**Review by uncenter — CHANGES_REQUESTED** [review]
Screenshot 2024-05-23 at 20 54 38 (Arc)

These pills/filters are unthemed.

Screenshot 2024-05-23 at 20 55 15 (Arc)

Chart elements are unthemed but are SVGs so completely themeable (`[fill="#XYZ"] { fill: @<color> !important; }`.

Screenshot 2024-05-23 at 20 56 08 (Arc)

Play icon and badge is unthemed.

Screenshot 2024-05-23 at 20 56 48 (Arc)

Logo is themeable, see https://github.com/catppuccin/userstyles/blob/main/docs/tips-and-tricks.md#img-elements.

**Comment by uncenter** [comment]
Click this button here to re-request a review once you have everything themed and fixed:
Screenshot 2024-05-23 at 21 44 32 (Arc)

File: styles/listenbrainz/catppuccin.user.css [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+        }
+
+        .navbar-logo > img {
+          @svg: escape(
+            '<?xml version="1.0" encoding="UTF-8"?>
+          <svg id="a" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 27 30">
+            <defs><style>.b{fill:@{mauve};}.c{fill:@{peach};}</style></defs>
+            <polygon class="b" points="13 1 1 8 1 22 13 29 13 1"/>
+            <polygon class="c" points="14 1 26 8 26 22 14 29 14 1"/>
+          </svg>'
```
Comment (uncenter) [inline styles/listenbrainz/catppuccin.user.css:100]:
```suggestion
            '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 27 30"><path fill="@{mauve}" d="M13 1 1 8v14l12 7z"/><path fill="@{peach}" d="m14 1 12 7v14l-12 7z"/></svg>'
```

File: styles/listenbrainz/catppuccin.user.css [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+    .musicbrainz-profile-button,
+    .lb-follow-button.block,
+    .pill.secondary.active {
+      color: @text;
+      background-color: fade(@accent-color, 30%);
+      border-color: transparent;
+      &:hover {
+        color: @text;
+        background-color: fade(@accent-color, 40%);
+      }
```
Comment (uncenter) [inline styles/listenbrainz/catppuccin.user.css:245]:
```suggestion
      color: @crust;
      background-color: @accent-color;
      border-color: transparent;

      &:hover {
        background-color: darken(@accent-color, 5%);
      }
```

**Review by uncenter — CHANGES_REQUESTED** [review]

File: styles/listenbrainz/catppuccin.user.css (hunk 2 above)
Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:245]: Hmm, this change looks a _lot_ nicer on dark flavours, which I like! However the button text becomes harder to read on Latte. Perhaps a separate style for Latte in particular might be needed. 🤔

Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css:238]: Huh really? How does it look on latte for you?

Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:238]: Not *awful*, but definitely less legible than the previous version to my eye.
Screenshot 2024-05-25 at 13 23 20
Screenshot 2024-05-25 at 13 23 54
I generally find light themes too bright anyway, though, so I may not be the best judge. 🤷‍♀️

Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css]: I tried slightly `lighten()`ing only the Latte version of crust, and I'm finding this a lot more legible. Otherwise basically as suggested. 👍
Screenshot 2024-05-27 at 10 13 27

**Comment by 00dani** [comment]
Hey, sorry for the holdup on these fixes! Should be all good now.

The clipped corners for missing or unlinked album art weren't technically a bug with this userstyle. It turns out vanilla ListenBrainz also clips those corners in the same way - it's just much harder to notice with the default colours. Nonetheless, it's fixed now.

I've also restored the missing shadows, making sure they look nice with both light and dark flavours. Those weren't exactly a mistake either, I initially removed the shadows rather than recolouring them intentionally because I liked the flatter look. However keeping them with appropriate recolouring is more consistent with vanilla ListenBrainz so. 👍

I did notice that since I last worked on this, ListenBrainz vanilla has introduced _another_ bug in their default styles: the search bar now has an identifying icon on it, but not in the correct place. With no userstyles applied:
Screenshot 2024-06-10 at 17 08 03

Should I fix that problem in the Catppuccin userstyle, or is it kind of "out of our jurisdiction" to fix upstream bugs?

**Comment by isabelroses** [comment]
> Should I fix that problem in the Catppuccin userstyle, or is it kind of "out of our jurisdiction" to fix upstream bugs?

We leave it up to them to fix it. Its much easier to manage the userstyles that way.

File: scripts/userstyles.yml [hunk 3]
```diff
@@ -519,7 +521,15 @@ userstyles:
...
-    current-maintainers: [*isabelroses]
+      current-maintainers: [*isabelroses]
+  listenbrainz:
+    name: ListenBrainz
+    categories: [music, entertainment]
+    icon: musicbrainz
+    color: maroon
+    readme:
+      app-link: 'https://listenbrainz.org'
+      current-maintainers: [*00dani]
```
Comment (isabelroses) [inline scripts/userstyles.yml:532]: We have recently hoisted this value up.
```suggestion
  listenbrainz:
    name: ListenBrainz
    categories: [music, entertainment]
    icon: musicbrainz
    color: maroon
    readme:
      app-link: 'https://listenbrainz.org'
    current-maintainers: [*00dani]
```

**Comment by isabelroses** [comment]
Your probaly going to have to merge with main such that https://github.com/catppuccin/userstyles/pull/908#discussion_r1633014580, won't cause a CI failure.

**Comment by 00dani** [comment]
It looks like this style's passing CI now, but the new go.dev userstyle has a problem:
```
$ deno task lint
Task lint deno run -A ./scripts/lint/main.ts
styles/go.dev/catppuccin.user.css:382
381│       #catppuccin(@lightFlavor, @accentColor);
382│     }
383│   }
   ╰─► error variable @svg is undefined

styles/go.dev/catppuccin.user.css:498:7
497│     .go-Main-navMobile label.go-Label::before {
498│       @svg:escape ('<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="@{text}"><path d="M0 0h24v24H0V0z" fill="none" opacity=".87"/><path d="M0 0h24v24H0V0z" fill="none"/><path d="M4 13c.55 0 1-.45 1-1s-.45-1-1-1-1 .45-1 1 .45 1 1 1zm0 4c.55 0 1-.45 1-1s-.45-1-1-1-1 .45-1 1 .45 1 1 1zm0-8c.55 0 1-.45 1-1s-.45-1-1-1-1 .45-1 1 .45 1 1 1zm3 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"/></svg>');
499│       background-image: url("data:image/svg+xml,@{svg}");
   ╰─► error Unexpected unknown at-rule "@svg:escape" (at-rule-no-unknown)
```
I think it's just that this particular SVG has been written as `@svg:escape (` rather than `@svg: escape(`, confusing the LESS parser?

**Comment by isabelroses** [comment]
> I think it's just that this particular SVG has been written as `@svg:escape (` rather than `@svg: escape(`, confusing the LESS parser?

Don't worry about this it was https://github.com/catppuccin/userstyles/pull/973 fixed here.

**Review by isabelroses — DISMISSED** [review]
Linters passing. I can't find any issues on the website. Happy to approve.

File: scripts/userstyles.yml [hunk 4]
```diff
@@ -531,6 +533,14 @@ userstyles:
     readme:
       app-link: "https://www.linkedin.com"
     current-maintainers: [*isabelroses]
+  listenbrainz:
+    name: ListenBrainz
+    categories: [music, entertainment]
+    icon: musicbrainz
+    color: maroon
```
Comment (uncenter) [inline scripts/userstyles.yml:540]: Where is `maroon` from? We should go with the dominant colors on the website itself, which would either be `peach` or `mauve` in this case I think.

File: styles/listenbrainz/catppuccin.user.css [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+    // We can't control which theme is passed to Toastify's props from CSS, but
+    // we can replace the default theme with whatever colours we want.
+    --toastify-color-light: @overlay0;
+    --toastify-text-color-light: @text;
+    --toastify-color-info: @blue;
+    --toastify-color-success: @green;
+    --toastify-color-warning: @yellow;
+    --toastify-color-error: @red;
+
+    @button-text: if(@lookup = latte, lighten(@crust, 10%), @crust);
```
Comment (uncenter) [inline styles/listenbrainz/catppuccin.user.css:81]: Can you elaborate on why this is needed again? I vaguely recall something about contrast? I think we can avoid that, the colors are okay as designed and we don't do this in any other userstyle.

File: styles/listenbrainz/catppuccin.user.css [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+    @button-bg: @accent-color;
```
Comment (uncenter) [inline styles/listenbrainz/catppuccin.user.css:82]: Let's avoid aliasing variables like this and simply use `accent-color` when needed.

File: styles/listenbrainz/catppuccin.user.css [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+      @flipped-peach: if(
+        @lookup = latte,
+        @catppuccin[@mocha][@peach],
+        @catppuccin[@latte][@peach]
+      );
+      @flipped-mauve: if(
+        @lookup = latte,
+        @catppuccin[@mocha][@mauve],
+        @catppuccin[@latte][@mauve]
+      );
```
Comment (uncenter) [inline styles/listenbrainz/catppuccin.user.css:609]: Why is this needed?

**Review by uncenter — CHANGES_REQUESTED** [review]

**Review by uncenter — CHANGES_REQUESTED** [review]
On the sitewide stats page (https://listenbrainz.org/statistics/) the text in these pills is unthemed:

CleanShot 2024-06-14 at 19 00 03@2x

File: styles/listenbrainz/catppuccin.user.css (hunk 7 above)
Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:609]: The [homepage](https://listenbrainz.org/?redirect=false) uses orange-to-purple gradients directly behind the text - using the matching versions of mauve and peach didn't give the text sufficient contrast to be legible. Flipping to the opposite scheme's version of the same colours seemed to do the trick.

I'm happy to try other approaches, of course.

File: scripts/userstyles.yml (hunk 4 above)
Comment (00dani) [reply scripts/userstyles.yml:540]: I have a vague memory of copying this info from the MusicBrainz section in userstyles.yml, keeping the same colour for consistency between the different MetaBrainz sites.

But there _isn't_ a MusicBrainz section in userstyles.yml, so I actually have no idea where `maroon` came from. I'll swap it out for `peach`. 🤷‍♀️

File: styles/listenbrainz/catppuccin.user.css (hunk 5 above)
Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:81]: I found `@crust` text on an `@accent-color` button very legible with dark flavours, but a little harder to make out on Latte. As I mentioned at the time though, I generally find light themes less legible anyway, so I'm probably not the best judge. Happy to just use unmodified `@crust`.

File: styles/listenbrainz/catppuccin.user.css [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+          }
+        }
+      }
+    }
+
+    .react-datetime-picker {
+      .react-datetime-picker__wrapper {
+        border-color: @mantle;
+      }
+      .react-calendar {
```
[the remark this answers is not in the document]
Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:471]: The border I just didn't think to theme. The yellow day I didn't spot at all, since that's "today" when you have some other day selected. Both fixed. 👍

File: styles/listenbrainz/preview.webp
[the remark this answers is not in the document]
Comment (00dani) [reply styles/listenbrainz/preview.webp:1]: The preview image needed re-capturing, since those sections have changed since I made it. Done!

File: styles/listenbrainz/catppuccin.user.css (hunk 5 above)
Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css]: Yep, the light theme is not technically WCAG compliant so the colors do need to be modified - but that isn't our responsibility, that will need to be changed in the upstream palette repository.

File: styles/listenbrainz/catppuccin.user.css (hunk 7 above)
Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css:616]: How does it look if you just use the colors as normal instead of the flipped versions?

Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:616]: The flipped versions look like this:
Screenshot 2024-06-16 at 11-58-38 ListenBrainz
Screenshot 2024-06-16 at 11-58-08 ListenBrainz

And the unflipped versions look like this:
Screenshot 2024-06-16 at 12-01-06 ListenBrainz
Screenshot 2024-06-16 at 12-01-42 ListenBrainz

Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css:616]: How do the unflipped versions look if you invert the text color?

Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:623]: The top half of the page looks fine:
Screenshot 2024-07-01 at 11-57-50 ListenBrainz
Screenshot 2024-07-01 at 11-58-16 ListenBrainz

But the bottom half has links that become illegible:
Screenshot 2024-07-01 at 11-58-51 ListenBrainz
Screenshot 2024-07-01 at 11-58-51 ListenBrainz

Colouring the links themselves differently would be an option?

**Review by isabelroses — APPROVED** [review]

File: styles/listenbrainz/catppuccin.user.css (hunk 7 above)
Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css:623]: Yep let's try that.

Comment (00dani) [reply styles/listenbrainz/catppuccin.user.css:623]: Since we want the links to still look like links, I tried using the blue and sky from the "flipped" flavour like I did with the background before, and that seems to look quite good (hovering over the second link to show how it looks on hover):
Screenshot 2024-07-05 at 13-57-26 ListenBrainz
Screenshot 2024-07-05 at 13-57-41 ListenBrainz

Are we okay to "flip" for just the link colours, keeping everything else the same, or is there a better approach I'm missing? 🤔

Comment (uncenter) [reply styles/listenbrainz/catppuccin.user.css:623]: Yes that sounds good.

**Review by uncenter — APPROVED** [review]

**Review by isabelroses — APPROVED** [review]
